django/db/models/fields/composite.py

```diff
@@ -0,0 +1,176 @@
+from collections.abc import Iterable
+
+from django.core.exceptions import FieldDoesNotExist
+from django.db.models import Field
+from django.db.models.expressions import Col, Expression
+from django.db.models.lookups import Exact, In
+from django.db.models.signals import class_prepared
+from django.utils.functional import cached_property
+
+
+class TupleExact(Exact):
+    def get_prep_lookup(self):
+        if not isinstance(self.lhs, Cols):
+            raise ValueError(
+                "The left-hand side of the 'exact' lookup must be an instance of Cols"
+            )
+        if not isinstance(self.rhs, Iterable):
+            raise ValueError(
+                "The right-hand side of the 'exact' lookup must be an iterable"
+            )
+        if len(list(self.lhs)) != len(list(self.rhs)):
+            raise ValueError(
+                "The left-hand side and right-hand side of the 'exact' lookup must "
+                "have the same number of elements"
+            )
+
+        return super().get_prep_lookup()
+
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, WhereNode
+
+        cols = self.lhs.get_source_expressions()
+        exprs = [Exact(col, val) for col, val in zip(cols, self.rhs)]
+
+        return compiler.compile(WhereNode(exprs, connector=AND))
+
+
+class TupleIn(In):
+    def get_prep_lookup(self):
+        if not isinstance(self.lhs, Cols):
+            raise ValueError(
+                "The left-hand side of the 'in' lookup must be an instance of Cols"
+            )
+        if not isinstance(self.rhs, Iterable):
+            raise ValueError(
+                "The right-hand side of the 'in' lookup must be an iterable"
+            )
+        if not all(isinstance(vals, Iterable) for vals in self.rhs):
+            raise ValueError(
+                "The right-hand side of the 'in' lookup must be an iterable of "
+                "iterables"
+            )
+        lhs_len = len(tuple(self.lhs))
+        if not all(lhs_len == len(tuple(vals)) for vals in self.rhs):
+            raise ValueError(
+                "The left-hand side and right-hand side of the 'in' lookup must "
+                "have the same number of elements"
+            )
+
+        return super().get_prep_lookup()
+
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, OR, WhereNode
+
+        exprs = []
+        cols = self.lhs.get_source_expressions()
+
+        for vals in self.rhs:
+            exprs.append(
+                WhereNode(
+                    [Exact(col, val) for col, val in zip(cols, vals)], connector=AND
+                )
+            )
+
+        return compiler.compile(WhereNode(exprs, connector=OR))
+
+
+class Cols(Expression):
```

Comment:
I wonder if there is an opportunity to merge this `TuplesIn`, `Cols`, and friends logic with `MultiColSource` so it's less of an 👽. They both do very similar thing.

Reply:
I merged `Cols` with `MultiColSource` 
[ad51da4](https://github.com/django/django/pull/18056/commits/ad51da49be839cde58d01e142be97d0750d5ceb9) however, I'm not sure it's correct.

As far as I understand, `MultiColSource` was meant to represent columns in a JOIN, and as such, it has a `sources` field. `Cols`, on the other hand, was meant to represent a list of columns and it doesn't need a `sources` field. WDYT?

Reply:
I reverted back to `Cols`. Please resolve if you agree.